game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProSortMoveOptionsUtils.java

```diff
@@ -125,195 +115,126 @@ private ProSortMoveOptionsUtils() {}
+              distance2 += data.getMap().getDistanceIgnoreEndForCondition(territory2, t, predicate);
+            }
+            if (distance1 != distance2) {
+              return distance1 - distance2;
             }
           }
-          return o1.getKey().getType().getName().compareTo(o2.getKey().getType().getName());
+
+          return unitType1.getName().compareTo(unitType2.getName());
         });
     final Map<Unit, Set<Territory>> sortedUnitAttackOptions = new LinkedHashMap<>();
     for (final Map.Entry<Unit, Set<Territory>> entry : list) {
       sortedUnitAttackOptions.put(entry.getKey(), entry.getValue());
     }
     return sortedUnitAttackOptions;
   }
+
+  private static Collection<Territory> removeWinningTerritories(
+      final Collection<Territory> territories,
+      final GamePlayer player,
+      final Map<Territory, ProTerritory> attackMap,
+      final ProOddsCalculator calc) {
+    return territories.stream()
+        .filter(
+            t -> {
+              final ProTerritory patd = attackMap.get(t);
+              if (patd.getBattleResult() == null) {
+                patd.estimateBattleResult(calc, player);
+              }
+              return !patd.isCurrentlyWins();
+            })
+        .collect(Collectors.toList());
+  }
+
+  private static double calculateAttackEfficiency(
+      final ProData proData,
+      final GamePlayer player,
+      final Map<Territory, ProTerritory> attackMap,
+      final Collection<Territory> territories,
+      final Unit unit) {
+    final GameData data = proData.getData();
+
+    int minPower = Integer.MAX_VALUE;
+    for (final Territory t : territories) {
+      final Collection<TerritoryEffect> effects = TerritoryEffectHelper.getEffects(t);
+      final UnitBattleComparator comparator =
+          new UnitBattleComparator(false, proData.getUnitValueMap(), effects, data);
+
+      final List<Unit> defendingUnits =
+          t.getUnitCollection().getMatches(Matches.enemyUnit(player, data));
+      final Collection<Unit> sortedUnits =
+          CollectionUtils.createSortedCollection(attackMap.get(t).getUnits(), comparator);
+      // Compare the difference in total power when including the unit or not.
+      int powerDifference = 0;
+      for (final boolean includeUnit : new boolean[] {false, true}) {
+        if (includeUnit) {
+          sortedUnits.add(unit);
+        }
+        powerDifference +=
+            (includeUnit ? 1 : -1)
+                * DiceRoll.getTotalPower(
+                    DiceRoll.getUnitPowerAndRollsForNormalBattles(
+                        sortedUnits,
+                        defendingUnits,
+                        sortedUnits,
+                        false,
+                        data,
+                        t,
+                        effects,
+                        false,
+                        null),
+                    data);
+      }
+      if (powerDifference < minPower) {
+        minPower = powerDifference;
+      }
```

Comment:
Unrelated to this PR, just a general thought.
I think writing this weird loop differently would make it more readable:
```java
final int powerWithout = DiceRoll.getTotalPower(
              DiceRoll.getUnitPowerAndRollsForNormalBattles(
                  sortedUnits,
                  defendingUnits,
                   sortedUnits,
                   false,
                   data,
                   t,
                   effects,
                   false,
                   null),
                 data);
sortedUnits.add(unit);
final int powerWith = DiceRoll.getTotalPower(
              DiceRoll.getUnitPowerAndRollsForNormalBattles(
                  sortedUnits,
                  defendingUnits,
                   sortedUnits,
                   false,
                   data,
                   t,
                   effects,
                   false,
                   null),
                 data);

minPower = Math.min(powerWith - powerWithout, minPower);
```